apps/hash-graph/lib/graph/src/store/postgres/knowledge/entity.rs

```diff
@@ -40,251 +39,267 @@ impl<C: AsClient> PostgresStore<C> {
-            let Some(intersected_temporal_axes) = temporal_axes.intersect_variable_interval(variable_interval) else {
-                // `traverse_entity` is called with the returned entities from `read` with
-                // `temporal_axes`. This implies, that the version interval of `entity` overlaps
-                // with `temporal_axes`. `variable_interval` returns `None` if there are
-                // no overlapping points, so this should never happen.
-                unreachable!("the version interval of the entity does not overlap with the variable axis's time interval");
-            };
-
-            if current_resolve_depths.is_of_type.outgoing > 0 {
-                let entity_type_id =
-                    OntologyTypeVertexId::from(entity.metadata.entity_type_id().clone());
-                subgraph.edges.insert(Edge::KnowledgeGraph {
-                    vertex_id: entity_vertex_id,
-                    outward_edge: KnowledgeGraphOutwardEdge::ToOntology(OutwardEdge {
-                        kind: SharedEdgeKind::IsOfType,
-                        reversed: false,
-                        right_endpoint: entity_type_id.clone(),
-                    }),
-                });
-
-                self.traverse_entity_type(
-                    &entity_type_id,
-                    traversal_context,
-                    subgraph,
-                    GraphResolveDepths {
-                        is_of_type: OutgoingEdgeResolveDepth {
-                            outgoing: current_resolve_depths.is_of_type.outgoing - 1,
-                            ..current_resolve_depths.is_of_type
-                        },
-                        ..current_resolve_depths
-                    },
-                    intersected_temporal_axes.clone(),
-                )
-                .await?;
-            }
+        graph_resolve_depths: GraphResolveDepths,
+        traversal_context: &mut TraversalContext,
+        subgraph: &mut Subgraph,
+    ) -> Result<(), QueryError> {
+        let time_axis = temporal_axes.variable_time_axis();
 
-            if current_resolve_depths.has_left_entity.incoming > 0 {
-                for outgoing_link_entity in <Self as Read<Entity>>::read(
-                    self,
-                    &Filter::for_outgoing_link_by_source_entity_id(entity_vertex_id.base_id),
-                    &intersected_temporal_axes,
-                )
-                .await?
-                {
-                    let link_entity_interval = outgoing_link_entity
-                        .metadata
-                        .temporal_versioning()
-                        .variable_time_interval(time_axis);
+        let mut queue = entity_vertex_ids
+            .into_iter()
+            .map(|id| (id, graph_resolve_depths, temporal_axes.clone()))
+            .collect::<Vec<_>>();
+
+        while !queue.is_empty() {
+            for (entity_vertex_id, graph_resolve_depths, temporal_axes) in mem::take(&mut queue) {
```

Comment:
What's the reason for the loop within the `while` rather than just doing this as part of the `while`?

I guess it could impact bounds checks?

While what I'm about to say is potentially a micro-optimisation, I also want to make sure we don't have this `for` loop for a different micro-optimisation as it seems strange to me.

Wouldn't it likely be faster to not call `mem::take` here and reuse capacity of the current queue? Otherwise we're allocating a new queue every time and have to wait for its grow logic whenever we append a certain amount to the queue (whereas if we started off with a big queue we'd likely not have to wait for reallocations)

Reply:
The memory cannot be reused here as we're pushing new items to the queue as we loop over the old queue.

Further down the road, I expect to remove the `for` loop here and pass a big part of the queue to the database layer directly. Mainly, I like to see the `queue` as "Items, I like to process at once" or at least as "Items, which _could_ be processed at once". Having everything in one big `for` loop does not help with this approach.

Reply:
> The memory cannot be reused here as we're pushing new items to the queue as we loop over the old queue.

I don't think that's true?
```rust
while !queue.is_empty() {
    let (entity_vertex_id, graph_resolve_depths, temporal_axes) = queue.pop().unwrap();
    ...

    queue.push((...));
}

Reply:
You're right. I added a comment to describe the reasoning: https://github.com/hashintel/hash/pull/2174/commits/b5af96050f865106cb226cc562b700efa4db10d0

Reply:
Thanks, makes sense that we're leaving the pattern in to be ready to process that whole queue in one go as a batch (rather than iteratively in a `for` loop)